### PR TITLE
Allow `T.attached_class` in modules

### LINKED ISSUE #2718: T.attached_class can't be used in modules that are `extend`ed into a class and become singleton methods
### Problem

I have a class method. It's defined inside of a module that uses `mixes_in_class_methods`:
```ruby
module Foo
  module ClassMethods
    sig {returns(T.self_type)}
    def load; end
  end
  mixes_in_class_methods(ClassMethods)
end

class Invoice
  include Foo
end
```

Sorbet thinks that `Invoice.load` returns `T.class_of(Invoice)`, but it actually returns `Invoice`.

### Proposed solution

I'd like to be able to use `T.attached_class`, but Sorbet doesn't realize that I'm in a singleton context. 

I'd like to add `class_methods!` as an annotation, similar to `interface!`:
```ruby
module Foo
  module ClassMethods
    # Sorbet now knows that this is in a singleton context
    class_methods!

    # You can correctly type the method:
    sig {returns(T.attached_class)}
    def load; end
  end
end
```

Sorbet would make it illegal to `include` the module. Instead, you must either `extend` it or use `mixes_in_class_methods`.

### PR DESCRIPTION
**Pull request by jez**
### Motivation

Fixes #2718

A common pattern in Ruby involves having modules define methods that take care of instantiating the object they get mixed into in devious ways.

This will allow us to write sigs for those things.

### Summary

The implementation is somewhat straightforward, powered by a rewriter.
It translates a new DSL into a type_member declaration, which allows using `T.attached_class`.

The ensuing complexity mostly just comes from error handling code that needs to special case some new error messages.

In fact, the whole feature to enforce that an `initializable!` module is only included via `extend` is powered by the existing logic to report errors for redeclaring type members. This is nice, because it means the correctness of both of these features is linked (and if there is a bug, fixing it in one fixes it for both).

### Test plan

See included automated tests.

### CODE REVIEW

File: core/types/calls.cc [hunk 1]
```diff
@@ -2968,22 +2968,31 @@ class Magic_attachedClass : public IntrinsicMethod {
...
-                auto singletonClass = self.data(gs)->lookupSingletonClass(gs);
-                if (singletonClass.exists()) {
-                    e.addErrorNote(
-                        "`{}` represents instances of a class; `{}` represents the corresponding singleton class",
-                        self.show(gs), singletonClass.show(gs));
+                auto initializable = core::Names::declareInitializable().show(gs);
+                if (selfData->isModule()) {
+                    e.setHeader("`{}` must be marked `{}` before module instance methods can use `{}`", self.show(gs),
+                                initializable, "T.attached_class");
+                    // TODO(jez) Autocorrect to insert `initializable!`
```
Comment (jez) [inline core/types/calls.cc:2982]: I plan on doing this in a future change, because there's some code I want to refactor first.

**Comment by jez** [comment]
I don’t want to tie this feature to requires_ancestor. As implemented, this feature comes with none of the downsides that requires_ancestor has.

**Comment by jez** [comment]
I think it’s also important for people to realize that this will turn their module into a generic, and having to write `extend T::Generic` underscores this.

there’s also the point that the type_member generated by the rewrite is allowed to have a bound, which would not be possible to infer 100% of the time from a requires_ancestor declaration (because the bound could be a union type).

File: website/docs/attached-class.md [hunk 2]
```diff
@@ -222,3 +222,129 @@ type suggests to the caller that `Child.make` will return a `Child` instance
...
+
+class ParentModel
+  extend T::Sig
+  extend FinderMethods
+end
+
+class ChildModel < ParentModel
+end
+
+parent = ChildModel.find('pa_123')
```
Comment (bdewater) [inline website/docs/attached-class.md:261]: Shouldn't this be `ParentModel`?

File: website/docs/attached-class.md [hunk 3]
```diff
@@ -222,3 +222,129 @@ type suggests to the caller that `Child.make` will return a `Child` instance
...
+end
+def find_and_log(findable, id)
+  instance = findable.find(id)
+  puts("Found #{instance}")
+  instance
+end
+
+parent = find_and_log(ParentModel, 'pa_123')
+T.reveal_type(parent) # => `ParentModel`
+child = find_and_log(ParentModel, 'pa_123')
```
Comment (bdewater) [inline website/docs/attached-class.md:318]: Similarly, `ChildModel`?

**Comment by jez** [comment]
Current dependencies on/for this PR:
* master
  * **PR #6883** <a href="https://app.graphite.dev/github/pr/sorbet/sorbet/6883" target="_blank">Graphite</a> 
    * **PR #6757** <a href="https://app.graphite.dev/github/pr/sorbet/sorbet/6757" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/sorbet/sorbet/6757?utm_source=stack-comment).

**Comment by chaffeqa** [comment]
this is very much needed in our case where we have a lot of `ActiveSupport::Concern` modules that build our models 👍

**Comment by jez** [comment]
@chaffeqa thanks for the enthusiasm, but please refrain from commenting on pull requests with such comments, so that contributors can focus on code review.

You're always welcome to share things like this on our [Sorbet Slack](https://sorbet.org/slack) instance, either in #discuss or #internals.

File: test/testdata/infer/generics/initialize_class_bound.rb [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  interface!
+
+  sig {abstract.returns(Integer)}
+  def foo; end
+
+  module Factory
+    extend T::Generic
+    interface!
+
+    has_attached_class!(:out) { {upper: Thing} }
```
Comment (froydnj) [inline test/testdata/infer/generics/initialize_class_bound.rb:15]: We don't seem to have any tests for `has_attached_class!` with bounds, but the enclosing module is used in a context that doesn't satisfy those bounds.  Can we add a testcase like that, please?  (I assume that the error messages will be somewhat sensible, but maybe not!)

**Review by froydnj — APPROVED** [review]
Minor comments only, since so much of Sorbet's existing machinery is doing the heavy lifting here.

File: resolver/GlobalPass.cc [hunk 5]
```diff
@@ -61,9 +71,29 @@ bool resolveTypeMember(core::GlobalState &gs, core::ClassOrModuleRef parent, cor
...
             // `type_member`, then the grandparent class's file could be edited and Sorbet wouldn't include the
             // grandchild class's file in the set of files to retypecheck.
-            e.setHeader("Type `{}` declared by parent `{}` must be re-declared in `{}`", name.show(gs), parent.show(gs),
-                        sym.show(gs));
-            e.addErrorLine(parentTypeMember.data(gs)->loc(), "`{}` declared in parent here", name.show(gs));
+            if (code == core::errors::Resolver::HasAttachedClassIncluded) {
+                auto hasAttachedClass = core::Names::declareHasAttachedClass().show(gs);
+                if (sym.data(gs)->isModule()) {
+                    e.setHeader("`{}` declared by parent `{}` must be re-declared in `{}`", hasAttachedClass,
+                                parent.show(gs), sym.show(gs));
```
[the remark this answers is not in the document]
Comment (jez) [reply resolver/GlobalPass.cc:78]: I think we use "declared" for error messages like "must be declared `abstract!`" or `final!`, which is what I was modeling this messaging off of.

image

I think that all names here are bad unfortunately.

[the remark this answers is not in the document]
Comment (jez) [reply resolver/GlobalPass.cc:78]: yeah we don't have it for other arms, but we should be able to do them all at once. That work is tracked in this issue:

https://github.com/sorbet/sorbet/issues/5240

[the remark this answers is not in the document]
Comment (froydnj) [reply resolver/GlobalPass.cc:78]: Yup, `abstract!` et al being "declarations" makes sense and the parallels with `has_attached_class!` WFM.

**Pull request merged**